koji: support multi-build composes

teg opened the pull request:
This introduces three new job types, and splits builds for koji into an init job, potentially multiple build jobs and a finalize job. This will allow us to build different images on different architectures, all part of the same compose. And it allows us to drop kerberos credentials from the composer node, and only keep that in the workers.

This shows a sample of a compose consisting of two images being processed by one worker:

```
$ journalctl -b -u osbuild-worker\@1
osbuild-worker[19657]: Running 'koji-init' job 39f1c1ab-9d13-4e25-b995-a529230c8c43
osbuild-worker[19657]: 2020/11/10 22:42:31 Job 39f1c1ab-9d13-4e25-b995-a529230c8c43 finished
osbuild-worker[19657]: Waiting for a new job...
osbuild-worker[19657]: Running 'osbuild-koji' job f38868d7-3313-43ba-8a45-20ac9a0c4376
osbuild-worker[19657]: 2020/11/10 22:45:36 Job f38868d7-3313-43ba-8a45-20ac9a0c4376 finished
osbuild-worker[19657]: Waiting for a new job...
osbuild-worker[19657]: Running 'osbuild-koji' job 0dc55e5a-63f4-40f3-ac80-b2fa7a742294
osbuild-worker[19657]: 2020/11/10 22:49:23 Job 0dc55e5a-63f4-40f3-ac80-b2fa7a742294 finished
osbuild-worker[19657]: Waiting for a new job...
osbuild-worker[19657]: Running 'koji-finalize' job f9c049a4-db90-4bfb-a8f2-0440ddce224c
osbuild-worker[19657]: 2020/11/10 22:50:17 Job f9c049a4-db90-4bfb-a8f2-0440ddce224c finished
```

The result n koji is:
```
$ koji buildinfo 1
BUILD: name-version-release [1]
State: COMPLETE
Built by: osbuild-krb
Source: 
Volume: DEFAULT
Task: none
Finished: Tue, 10 Nov 2020 23:29:35 UTC
Tags: 
Extra: {'typeinfo': {'image': {}}}
Image archives:
/mnt/koji/packages/name/version/release/images/name-version-release.x86_64.raw
/mnt/koji/packages/name/version/release/images/name-version-release.x86_64.qcow2
```

ondrejbudai commented on internal/jobqueue/jobqueue.go:
```diff
@@ -56,7 +56,7 @@ type JobQueue interface {
 	// finished, respectively.
 	//
 	// If the job is finished, its result will be returned in `result`.
-	JobStatus(id uuid.UUID, result interface{}) (queued, started, finished time.Time, canceled bool, err error)
```
A note about this function returning also the deps would be nice.

teg replied:
Fixed.

ondrejbudai commented on internal/kojiapi/api/openapi.yml:
```diff
@@ -112,7 +112,7 @@ components:
           items:
             $ref: '#/components/schemas/ImageStatus'
-        koji_task_id:
+        koji_build_id:
```
@gicmo This change may interfere with `koji-osbuild`.

gicmo replied:
FTR, indeed: https://github.com/osbuild/koji-osbuild/blob/f529ffe394f370e0ffde88d98f9ce038cbfd7937/plugins/builder/osbuild.py#L116

teg replied:
I reverted this. Thanks @ondrejbudai, and sorry for dropping the ball on this!

ondrejbudai reviewed (changes requested):
This looks... good? ...huge? ...slightly scary? I think this is most importantly AWESOME! It really moves us to be able to finish the Koji integration. I have some notes:

1) The access to `dynamicArgs` by a numeric index seems a bit fragile to me. I'm not sure how to solve it in a more robust way though.

2) We need more tests for this I think. This adds a lot of non-trivial code that would really benefit from better coverage.

`< len(deps)` should be solved before merging this, other comments can be resolved in follow-ups I think.

teg commented on cmd/osbuild-composer/config_test.go:
```diff
@@ -30,12 +29,6 @@ func TestConfig(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, config)
 
-	server, ok := config.Koji.Servers["example.com"]
```
Done!

teg commented on internal/kojiapi/server.go:
```diff
@@ -248,15 +261,38 @@ func (h *apiHandlers) GetComposeId(ctx echo.Context, idstr string) error {
+
+	var buildResults []worker.OSBuildKojiJobResult
+	var imageStatuses []api.ImageStatus
+	for i := 1; i < len(deps)-1; i++ {
```
Oops! Thanks! Fixed.

teg commented on cmd/osbuild-worker/jobimpl-koji-finalize.go:
```diff
@@ -0,0 +1,180 @@
+			return err
+		}
+		if !buildArgs.OSBuildOutput.Success || buildArgs.KojiError != nil {
+			failure = true
```
Indeed! Done.

teg commented:
@ondrejbudai I added some more tests and addressed the issue you found (among others). I agree that indexing the dynamic args array is not ideal, happy for that to be reworked to be safer. Also hate the name ;)

If Schutzbot is happy I think that can all be addressed in follow-ups though.

ondrejbudai reviewed (approved):
Everything seems to be OK now. I did some manual testing just to be extra-sure and I haven't found any regressions.

larskarlitski reviewed (dismissed):
Amazing! ❤️ 

Would love to run this against `koi-osbuild`'s test once before making the release, just to make sure.

teg commented:
@larskarlitski ran the kojji-osbuild test suite on this PR (we should have reverse dep test here), and the only failing test in schutzbot was a failed download during base tests that succeeded on a previous run. This PR does not test the base tests, so this should be good to go.

larskarlitski commented:
> @larskarlitski ran the kojji-osbuild test suite on this PR (we should have reverse dep test here), and the only failing test in schutzbot was a failed download during base tests that succeeded on a previous run. This PR does not test the base tests, so this should be good to go.

Indeed! All looking good. Thanks for this ❤️
